### PR TITLE
[WIP] Resolve numpy 1.25 deprecations and issues

### PR DESCRIPTION
**Pull request by bryevdv**
This PR fixes two issues that cropped up with recent Numpy 1.25:

* A new class revealed an unprotected usage of `._cunumeric` that could file. For now, resolve by using safe `is_implemented`. 
* Mypy errors due to the deprecation of `find_common_type`

The latter changes have resulting in a few test failures. Still trying to apply the somewhat obscure advice for https://github.com/numpy/numpy/releases/tag/v1.25.0 

Alternatively we could simply ignore the new mypy errors and continue using `find_common_type` for now.

### CODE REVIEW

**Comment by bryevdv** [comment]
@seberg I simplified in https://github.com/nv-legate/cunumeric/pull/972/commits/622ae6d733b468e2d17a9cad4409ac64e59a3e05

Still seeing local failures e.g. 

<details>

```
dev310 ❯ LEGATE_TEST=1 REALM_SYNTHETIC_CORE_MAP= legate --omps 2 --ompthreads 1 --numamem 0 /home/bryan/work/cunumeric/tests/integration/test_binary_ufunc.py

[0 - 7f8d23f0d000]    0.000114 {4}{threads}: reservation ('CPU proc 1d00000000000003') cannot be satisfied
============================================================================================ test session starts =============================================================================================
platform linux -- Python 3.10.11, pytest-7.3.2, pluggy-1.0.0
rootdir: /home/bryan/work/cunumeric
configfile: pyproject.toml
plugins: mock-3.11.1, cov-4.1.0, lazy-fixture-0.6.3
collected 1 item                                                                                                                                                                                             

tests/integration/test_binary_ufunc.py F                                                                                                                                                               [100%]

================================================================================================== FAILURES ==================================================================================================
__________________________________________________________________________________________________ test_all __________________________________________________________________________________________________

    def test_all():
        # TODO: right now we will simply check if the operations work
        # for some boring inputs. For some of these, we will want to
        # test corner cases in the future.
    
        # Math operations
        ops = [
            "add",
            # "divmod",
            "equal",
            "fmax",
            "fmin",
            "greater",
            "greater_equal",
            # "heaviside",
            # "ldexp",
            "less",
            "less_equal",
            "logical_and",
            "logical_or",
            "logical_xor",
            "maximum",
            "minimum",
            "multiply",
            "not_equal",
            "subtract",
            "true_divide",
        ]
    
        # We want to test array-array, array-scalar, and scalar-array cases
        arrs = (
            np.random.randint(3, 10, size=(4, 5)).astype("I"),
            np.random.uniform(size=(4, 5)).astype("F"),
        )
        scalars = (
            np.uint64(2),
            np.int64(-3),
            np.random.randn(1)[0],
            np.complex64(1 + 1j),
        )
    
        for arr1, arr2 in product(arrs, arrs):
            check_ops(ops, (arr1, arr2))
    
        for arr, scalar in product(arrs, scalars):
>           check_ops(ops, (arr, scalar))

tests/integration/test_binary_ufunc.py:117: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/integration/test_binary_ufunc.py:55: in check_ops
    check_result(op, in_np, out_np, out_num)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

op = 'add', in_np = (array([[5, 8, 8, 4, 7],
       [4, 9, 9, 8, 5],
       [8, 4, 9, 4, 6],
       [9, 4, 6, 9, 3]], dtype=uint32), 2)
out_np = array([[ 7, 10, 10,  6,  9],
       [ 6, 11, 11, 10,  7],
       [10,  6, 11,  6,  8],
       [11,  6,  8, 11,  5]], dtype=uint32)
out_num = array([[ 7, 10, 10,  6,  9],
       [ 6, 11, 11, 10,  7],
       [10,  6, 11,  6,  8],
       [11,  6,  8, 11,  5]], dtype=uint64)

    def check_result(op, in_np, out_np, out_num):
        result = allclose(out_np, out_num) and out_np.dtype == out_num.dtype
        if not result:
            print(f"cunumeric.{op} failed the test")
            print("Inputs:")
            for arr in in_np:
                print(arr)
                print(f"dtype: {arr.dtype}")
            print("NumPy output:")
            print(out_np)
            print(f"dtype: {out_np.dtype}")
            print("cuNumeric output:")
            print(out_num)
            print(f"dtype: {out_num.dtype}")
>           assert False
E           assert False

tests/integration/test_binary_ufunc.py:40: AssertionError
-------------------------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------------------------
allclose: different dtype: uint32 vs uint64
cunumeric.add failed the test
Inputs:
[[5 8 8 4 7]
 [4 9 9 8 5]
 [8 4 9 4 6]
 [9 4 6 9 3]]
dtype: uint32
2
dtype: uint64
NumPy output:
[[ 7 10 10  6  9]
 [ 6 11 11 10  7]
 [10  6 11  6  8]
 [11  6  8 11  5]]
dtype: uint32
cuNumeric output:
[[ 7 10 10  6  9]
 [ 6 11 11 10  7]
 [10  6 11  6  8]
 [11  6  8 11  5]]
dtype: uint64
============================================================================================== warnings summary ==============================================================================================
tests/integration/test_binary_ufunc.py::test_all
  /home/bryan/work/cunumeric/cunumeric/array.py:3925: RuntimeWarning: cuNumeric has not implemented result_type and is falling back to canonical NumPy. You may notice significantly decreased performance for this function call.
    return np.result_type(*args)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================================================================================== short test summary info ===========================================================================================
FAILED tests/integration/test_binary_ufunc.py::test_all - assert False
======================================================================================== 1 failed, 1 warning in 1.00s ========================================================================================
```

</details>

File: cunumeric/_ufunc/ufunc.py [hunk 1]
```diff
@@ -571,19 +571,20 @@ def _find_common_type(
...
+        # int, float, or complex in scalar_types."
+        types = []
         for arr, orig_arg in zip(arrs, orig_args):
             if arr.ndim == 0:
-                scalar_types.append(
-                    np.dtype(np.min_scalar_type(orig_arg))
-                    if use_min_scalar
-                    else arr.dtype
+                types.append(
+                    np.dtype(orig_arg) if use_min_scalar else arr.dtype
```
Comment (seberg) [inline cunumeric/_ufunc/ufunc.py:582]:
```suggestion
                    orig_arg if use_min_scalar else arr.dtype
```
sorry, lets try this.  My thought was to keep the `.type(0)` on the first branch initially, but it will be wrong without NEP 50 finalized.  You definitely need the scalar though.
(This will call `np.asarray(orig_arg)` effectively, but the old code did too)

Using `orig_arg` for any 0-D input should mean you can skip the `use_min_scalar` logic (but that is the follow up once this works.)

Comment (bryevdv) [reply cunumeric/_ufunc/ufunc.py:582]: Still seeing failures, all see to be type mismatches

File: cunumeric/_ufunc/ufunc.py [hunk 2]
```diff
@@ -571,19 +571,20 @@ def _find_common_type(
         )
         use_min_scalar = not (all_scalars or all_arrays or lossy_conversion)
```
Comment (seberg) [inline cunumeric/_ufunc/ufunc.py:572]:
```suggestion
        use_min_scalar = not (all_scalars or all_arrays)
```
Aha, I missed that the old code had this oddity; probably because `np.find_common_type` was weird...  It is not always correct this way and `result_type` applies the correct logic.
(The only reason this ever worked must be because `find_common_dtype` didn't even match how numpy does promotion, apparently...)

Anyway, removing this, the eager tests seem to passed locally for me.  (The `use_min_scalar` logic could be removed, always taking the `True` path is correct with `result_type`)

**Comment by bryevdv** [comment]
Thanks @seberg I applied your suggestions to #973 and things seem to work locally for me as well, I will close this PR in favor of that one
